rust/parquet/src/arrow/record_reader.rs

```diff
@@ -0,0 +1,674 @@
+use crate::data_type::DataType;
+use crate::errors::{ParquetError, Result};
+use crate::schema::types::ColumnDescPtr;
+use arrow::bitmap::Bitmap;
+use arrow::buffer::{Buffer, MutableBuffer};
+use arrow::builder::{BooleanBufferBuilder, BufferBuilderTrait};
+
+const MIN_BATCH_SIZE: usize = 1024;
+
+/// A `RecordReader` is a stateful column reader that delimits semantic records.
+pub struct RecordReader<T: DataType> {
+    column_desc: ColumnDescPtr,
+
+    records: MutableBuffer,
+    def_levels: Option<MutableBuffer>,
+    rep_levels: Option<MutableBuffer>,
+    null_bitmap: Option<BooleanBufferBuilder>,
+    column_reader: Option<ColumnReaderImpl<T>>,
+
+    /// Number of records accumulated in records
+    records_num: usize,
+
+    /// Last offset of `records_num` records
+    values_pos: usize,
+    values_seen: usize,
+    /// Starts from 1, number of values have been written to buffer
+    values_written: usize,
+    in_middle_of_record: bool,
+}
+
+struct FatPtr<T> {
+    ptr: *const T,
+    len: usize,
+}
+
+impl<T> FatPtr<T> {
+    fn new(ptr: *const T, len: usize) -> Self {
+        Self { ptr, len }
+    }
+
+    unsafe fn to_slice(&self) -> &[T] {
+        slice::from_raw_parts(self.ptr, self.len)
+    }
+
+    unsafe fn to_slice_mut(&self) -> &mut [T] {
+        slice::from_raw_parts_mut(self.ptr as *mut T, self.len)
+    }
+}
+
+impl<T: DataType> RecordReader<T> {
+    pub fn new(column_schema: ColumnDescPtr) -> Self {
+        let (def_levels, nullity_bitmap) = if column_schema.max_def_level() > 0 {
+            (
+                Some(MutableBuffer::new(MIN_BATCH_SIZE)),
+                Some(BooleanBufferBuilder::new(MIN_BATCH_SIZE)),
+            )
+        } else {
+            (None, None)
+        };
+
+        let rep_levels = if column_schema.max_rep_level() > 0 {
+            Some(MutableBuffer::new(MIN_BATCH_SIZE))
+        } else {
+            None
+        };
+
+        Self {
+            records: MutableBuffer::new(MIN_BATCH_SIZE),
+            def_levels,
+            rep_levels,
+            null_bitmap: nullity_bitmap,
+            column_reader: None,
+            column_desc: column_schema,
+            records_num: 0,
+            values_pos: 0,
+            values_seen: 0,
+            values_written: 0,
+            in_middle_of_record: false,
+        }
+    }
+
+    /// Set the current page reader.
+    pub fn set_page_reader(&mut self, page_reader: Box<PageReader>) -> Result<()> {
+        self.column_reader =
+            Some(ColumnReaderImpl::new(self.column_desc.clone(), page_reader));
+        Ok(())
+    }
+
+    /// Try to read `num_records` of column data into internal buffer.
+    ///
+    /// # Returns
+    ///
+    /// Number of actual records read.
+    pub fn read_records(&mut self, num_records: usize) -> Result<usize> {
+        let mut records_read = 0usize;
+
+        // end_of_column is used to mark whether we have reached the end of current page
```

Comment:
`current page reader` -> `current column chunk`? also `end_of_column` is kind of redundant, change it to `Used to ...`?